File: backend/app/auth/permissions.py
"""
Pure permission helpers shared by runtime auth code and tests.
"""

POWER_ADMIN_ONLY_ROLES = {"Admin", "Officer"}


def has_global_role(roles: list[dict], *role_names: str) -> bool:
    return any(
        r.get("role") in role_names and r.get("scope") == "global"
        for r in roles
    )


def has_tenant_role(roles: list[dict], tenant_id, *role_names: str) -> bool:
    tid = str(tenant_id)
    return any(
        r.get("role") in role_names and r.get("tenant_id") == tid
        for r in roles
    )


def is_reader(roles: list[dict], tenant_id) -> bool:
    return (
        has_tenant_role(roles, tenant_id, "Reader", "Writer", "PowerUser", "Approver", "Officer")
        or has_global_role(roles, "Auditor")
    )


def is_writer(roles: list[dict], tenant_id) -> bool:
    return has_tenant_role(roles, tenant_id, "Writer", "PowerUser")


def is_power_user(roles: list[dict], tenant_id) -> bool:
    return has_tenant_role(roles, tenant_id, "PowerUser")


def is_approver(roles: list[dict], tenant_id) -> bool:
    return has_tenant_role(roles, tenant_id, "Approver")


def is_officer(roles: list[dict], tenant_id) -> bool:
    return has_tenant_role(roles, tenant_id, "Officer")


def is_admin(roles: list[dict], tenant_id) -> bool:
    return has_tenant_role(roles, tenant_id, "Admin")


def is_auditor(roles: list[dict]) -> bool:
    return has_global_role(roles, "Auditor")


def is_power_admin(roles: list[dict]) -> bool:
    return has_global_role(roles, "PowerAdmin")

File: frontend/src/utils/permissions.js
export function hasGlobalRole(roles, ...names) {
  return roles.some(r => names.includes(r.role) && r.scope === 'global');
}

export function hasTenantRole(roles, tenantId, ...names) {
  return roles.some(r => names.includes(r.role) && r.tenant_id === tenantId);
}

export function canReadBookings(roles, tenantId) {
  return (
    hasTenantRole(roles, tenantId, 'Reader', 'Writer', 'PowerUser', 'Approver', 'Officer') ||
    hasGlobalRole(roles, 'Auditor')
  );
}

export function canWriteBookings(roles, tenantId) {
  return hasTenantRole(roles, tenantId, 'Writer', 'PowerUser');
}

export function canApproveBookings(roles, tenantId) {
  return hasTenantRole(roles, tenantId, 'Approver');
}

export function canManageRoles(roles, tenantId) {
  return hasTenantRole(roles, tenantId, 'Admin') || hasGlobalRole(roles, 'PowerAdmin');
}

export function canReadAccounts(roles, tenantId) {
  return canReadBookings(roles, tenantId) ||
    hasTenantRole(roles, tenantId, 'Admin') ||
    hasGlobalRole(roles, 'PowerAdmin');
}

export function canWriteAccounts(roles, tenantId) {
  // Admin manages role assignments only and has no account write access;
  // mirrors require_account_write and the accounts RLS write policy.
  return hasTenantRole(roles, tenantId, 'PowerUser') || hasGlobalRole(roles, 'PowerAdmin');
}

export function canPostJournalEntry(roles, tenantId) {
  return hasTenantRole(roles, tenantId, 'PowerUser');
}

export function canShowUserRoleRoute(roles) {
  return hasGlobalRole(roles, 'PowerAdmin') || roles.some(r => r.role === 'Admin');
}

export function canShowTenantsRoute(roles) {
  return hasGlobalRole(roles, 'PowerAdmin');
}

export function canShowAuditRoute(roles) {
  return hasGlobalRole(roles, 'Auditor', 'PowerAdmin') || roles.some(r => r.role === 'Officer');
}
